src/mlpack/methods/ann/activation_functions/sqnl_function.hpp

```diff
@@ -0,0 +1,153 @@
+/**
+ * @file sqnl_function.hpp
+ * @author Prince Gupta
+ *
+ * Definition and implementation of the SQNL(Square Non Linearlity)
+ * activation function.
+ *
+ * For more information, see the following paper.
+ *
+ * @code
+ * @misc{
+ *   author = {Adedamola Wuraola, Nitish Patel},
+ *   title = {SQNL: A New Computationally Efficient Activation Function},
+ *   year = {2018}
+ * }
+ * @endcode
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ANN_ACTIVATION_FUNCTIONS_SQNL_FUNCTION_HPP
+#define MLPACK_METHODS_ANN_ACTIVATION_FUNCTIONS_SQNL_FUNCTION_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artifical Neural Network. */ {
+
+/**
+ * The SQNL(Square Non Linearity) activation function, defined by
+ * 
+ * @f{eqnarray*}{
+ * f(x) &=& 1, if x > 2,
+ *          x + x^{2}/4, if -2 <= x < 0,
+ *          x - x^{2}/4, if 0 <= x <= 2,
+ *          -1, if x < -2
+ * f'(x) &=& 0, if x > 2 or x < -2,
+ *           1 - x / 2, if 0 <= x <= 2,
+ *           1 + x / 2, if -2 <= x < 0
+ * @f}
+ */
+class SQNLFunction
+{
+ public:
+  /**
+   * Computes the SQNL(Square Non Linearity) function.
+   * 
+   * @param x Input data.
+   * @return f(x).
+   */
+  static double Fn(const double x)
+  {
+      if (x > 2)
+        return 1;
+      else if (x >= 0 && x <= 2)
+        return x - std::pow(x, 2) / 4;
+      else if (x >= -2 && x < 0)
+        return x + std::pow(x, 2) / 4;
+      else
+        return -1;
+  }
+
+  /**
+   * Computes the SQNL(Square Non Linearity) function.
+   * 
+   * @param x Input data.
+   * @param y The resulting output activation.
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Fn(const InputVecType& x, OutputVecType& y)
+  {
+    y.set_size(arma::size(x));
+    for (size_t i = 0; i < x.n_elem; i++)
+    {
+      y(i) = Fn(x(i));
+    }
+  }
+
+  /**
+   * Computes the inverse of SQNL function y = f(x).
+   * Since y = 1 for all x >= 2 and y = -1 for all x <= -2 it returns values
+   * 2 and -2 for y == 1 and y == -1 respectively.
+   * Since y can only take values in range [-1,1] if input is some value out of bounds,
+   * value returned is 0.
+   * 
+   * @param y Input Data.
+   * @return f^{-1}(y).
+   */
+  static double Inv(const double y)
+  {
+    if (y >= 0 && y <= 1)
+      return 2 * (1 - std::sqrt(1 - y));
```

Comment:
Do you have a reference for this? For example the derivative of `x + x^2 / 4` is `1 + (x/2)` but I'm not sure can see where your formula comes from, maybe I missed something?

Reply:
This is for calculating inverse. Since SQNL function gives `y = 1 for all x >2` and `y = -1 for all x < -2`, during calculating inverse I set it up so it returns 2 for y= 1 and -2 for y  = -1. This way In `Deriv()` function I don't need to use `if` and `else` and simply the formula for derviative becomes 1 - abs(x/2) , it satisfies both `x-x^2 /4 ( 0<=x<=2)` and `x + x^2/4(-2<=x<0)` and x comes from `Inv(y)`.

Reply:
Not sure, but my guess is `std::sqrt(1 - y)` is more expensive compared with `1 + (x/2)`, I could be wrong.

Reply:
Yes it certainly is more expensive but without calculating inverse of y for SQNL function I couldn't find a way to calculate derivative that is 1 + x/2, and to calculate x we need to carry out `std::sqrt(1 - y)`. I may be wrong but I don't think we can get derivative directly from y like we can in sigmoid ( `y(1-y)` )

Reply:
Not sure what you mean with "I couldn't find a way to calculate derivative that is 1 + x/2", it's the derivative of x + x^2 / 4. Maybe I'm missing something?

Reply:
Any update on this one?

Reply:
It could be that my understanding of Backward() function is wrong somewhere but I've looked at it many times and:
Backward() Function in FFN does:
```
    boost::apply_visitor(BackwardVisitor(boost::apply_visitor(
        outputParameterVisitor, network[network.size() - i]),
        boost::apply_visitor(deltaVisitor, network[network.size() - i + 1]),
        boost::apply_visitor(deltaVisitor, network[network.size() - i])),
        network[network.size() - i]);
```
which means for layer l, parameter passed to its Backward function are:
outputParameter of layer l, delta of layer l + 1, and delta of layer l.

and Backward() function of BaseLayer (where SQNL is being used) is:
```  
void Backward(const arma::Mat<eT>& input,
                const arma::Mat<eT>& gy,
                arma::Mat<eT>& g)
  {
    arma::Mat<eT> derivative;
    ActivationFunction::Deriv(input, derivative);
    g = gy % derivative;
  }
```
which means the output of SQNL Layer is being passed to it when backward function is called.
Hence, for y = f(x), we have y as input during Backpropagation, so we can't output ` 1 + input/2`, since derivative is `1 +x/2` and not `1 + y/2` so, we'll have to calculate the Inverse() i.e x = f{^-1}(input) then return `1 + x / 2`.

> Not sure, but my guess is `std:: sqrt(1 - y)` is more expensive compared with `1 + (x/2)`, I could be wrong.

Since we have y as the input and not x we'll have to calculate the inverse first by `std:: sqrt(1 - y)`.

Sorry for the long answer. Please correct me if I am wrong (which I think I am somewhere)